File: 0938-range-sum-of-bst/0938-range-sum-of-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
       if (root == nullptr) {
            return 0;
        }

        int sum = 0;
        stack<TreeNode*> nodeStack;
        nodeStack.push(root);

        while (!nodeStack.empty()) {
            TreeNode* node = nodeStack.top();
            nodeStack.pop();

            if (node != nullptr) {
                if (node->val > low) {
                    nodeStack.push(node->left);
                }
                if (node->val < high) {
                    nodeStack.push(node->right);
                }
                if (low <= node->val && node->val <= high) {
                    sum += node->val;
                }
            }
        }

        return sum;
    }
};

File: 0938-range-sum-of-bst/0938-range-sum-of-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def rangeSumBST(self, root, low, high):
        """
        :type root: TreeNode
        :type low: int
        :type high: int
        :rtype: int
        """
        if root is None:
            return 0
        
        value = root.val if low <= root.val <= high else 0
        
        #putting "self" ensures recursive calls within the same class
        leftVal = self.rangeSumBST(root.left, low, high)
        rightVal = self.rangeSumBST(root.right, low, high)
        
        return value + leftVal + rightVal